src/compiler/nir/nir_opt_load_store_vectorize.c

```diff
@@ -0,0 +1,1029 @@
+         return false;
+   }
+
+   if (first->store_value)
+      vectorize_adjacent_stores(&b, low, high, first, second, new_bit_size, new_num_components, diff * 8u);
+   else
+      vectorize_adjacent_loads(&b, low, high, first, second, new_bit_size, new_num_components, diff * 8u);
+
+   return true;
+}
+
+static bool
+handle_barrier(struct state *state, nir_instr *instr)
+{
+   unsigned modes = 0;
+   if (instr->type == nir_instr_type_intrinsic) {
+      nir_intrinsic_instr *intrin = nir_instr_as_intrinsic(instr);
+      switch (intrin->intrinsic) {
+      case nir_intrinsic_barrier:
+      case nir_intrinsic_group_memory_barrier:
+      case nir_intrinsic_memory_barrier:
+      /* prevent speculative loads/stores */
+      case nir_intrinsic_discard_if:
+      case nir_intrinsic_discard:
+         modes = nir_var_all;
+         break;
+      case nir_intrinsic_memory_barrier_buffer:
+         modes = nir_var_mem_ssbo;
+         break;
+      case nir_intrinsic_memory_barrier_shared:
+         modes = nir_var_mem_shared;
+         break;
+      default:
+         return false;
+      }
+   } else if (instr->type == nir_instr_type_call) {
+      modes = nir_var_all;
+   } else {
+      return false;
+   }
+
+   while (modes) {
+      unsigned mode = u_bit_scan(&modes);
+      list_for_each_entry_safe(struct entry, entry, &state->entries[mode], head)
+         entry->behind_barrier = true;
+   }
+
+   return true;
+}
+
+/* false means they might be either different or same */
+static bool
+resources_different(nir_ssa_def *a, nir_ssa_def *b)
+{
+   if (!a || !b)
+      return false;
+
+   nir_const_value *acv = nir_src_as_const_value(nir_src_for_ssa(a));
+   nir_const_value *bcv = nir_src_as_const_value(nir_src_for_ssa(b));
+   if (acv && bcv)
+      return acv->u32[0] != bcv->u32[0];
+   else if (acv || bcv)
+      return false;
+
+   if (a->parent_instr->type == nir_instr_type_intrinsic &&
+       b->parent_instr->type == nir_instr_type_intrinsic) {
+      nir_intrinsic_instr *aintrin = nir_instr_as_intrinsic(a->parent_instr);
+      nir_intrinsic_instr *bintrin = nir_instr_as_intrinsic(b->parent_instr);
+      if (aintrin->intrinsic == nir_intrinsic_vulkan_resource_index &&
+          bintrin->intrinsic == nir_intrinsic_vulkan_resource_index) {
+         return nir_intrinsic_desc_set(aintrin) != nir_intrinsic_desc_set(bintrin) ||
+                nir_intrinsic_binding(aintrin) != nir_intrinsic_binding(aintrin) ||
+                resources_different(aintrin->src[0].ssa, bintrin->src[0].ssa);
+      }
+   }
+
+   return false;
+}
+
+static bool
+may_alias(struct entry *a, struct entry *b)
+{
+   assert(get_variable_mode(a) == get_variable_mode(b));
+
+   if ((a->key.var != b->key.var || resources_different(a->key.resource, b->key.resource)) &&
+       (a->access & ACCESS_RESTRICT) && (b->access & ACCESS_RESTRICT))
+      return false;
+   if (a->key.var != b->key.var || a->key.resource != b->key.resource)
```

Comment:
shouldn't this be == ?

Reply:
The idea here is that if it they have potentially different resources/variables which may alias, then they can't be compared. So, instead of comparing the offsets, it simply returns "true".

Changing it to "a->key.var == b->key.var && a->key.resource == b->key.resource" would have it return "true" without actually comparing the offsets (which it can, since they are definitely the same resource/variable).